### PR TITLE
Finish user edit, update, index, and destroy actions

### PR DESCRIPTION
**Pull request by llabake**
-  Update user
- Seed users
- Create an admin user
- Delete user

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -7,8 +7,7 @@ class User < ApplicationRecord
             format: { with: VALID_EMAIL_REGEX },
             uniqueness: { case_sensitive: false }
   has_secure_password
-  validates :password, presence: true, length: { minimum: 6 }
-
+  validates :password, presence: true, length: { minimum: 6 }, allow_nil: true
```
Comment (chalchuck) [inline app/models/user.rb]: why would you have password as blank?

File: app/models/user.rb [hunk 2]
```diff
@@ -7,8 +7,7 @@ class User < ApplicationRecord
...
   # Returns the hash digest of the given string.
   def User.digest(string)
```
Comment (chalchuck) [inline app/models/user.rb:12]: this method is wrongly defined. (This is not how we define methods in Ruby)

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,11 +1,20 @@
...
     @user = User.new
   end
 
   def show
     @user = User.find(params[:id])
   end
+
   def create
     @user = User.new(user_params)
     if @user.save
```
Comment (chalchuck) [inline app/controllers/users_controller.rb:20]: what happens when a user is saved??

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -16,10 +25,45 @@ def create
...
+  def update
+    if @user.update_attributes(user_params)
+      redirect_to @user, notice: 'Profile updated'
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
+    User.find(params[:id]).destroy
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: move this to a `before_action`

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -16,10 +25,45 @@ def create
...
+
+  def update
+    if @user.update_attributes(user_params)
+      redirect_to @user, notice: 'Profile updated'
+    else
+      render 'edit'
+    end
+  end
+
+  def destroy
```
Comment (chalchuck) [inline app/controllers/users_controller.rb:39]: add a conditional to ensure that a user can not delete herself/himself.

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -16,10 +25,45 @@ def create
...
+    end
+  end
+
+  def correct_user
+    @user = User.find(params[:id])
+    redirect_to(root_url) unless current_user?(@user)
+  end
+
+  def admin_user
+    redirect_to(root_url) unless current_user.admin?
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: Add a flash message indicating that this is only an ADMIN function.

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -16,10 +25,45 @@ def create
...
+  def logged_in_user
+    store_location
+    unless logged_in?
+      flash[:danger] = "Please log in."
+      redirect_to login_url
+    end
+  end
+
+  def correct_user
+    @user = User.find(params[:id])
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: Move this finder to a `before_action` you are re-using it in more than one place.

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -6,7 +6,7 @@ def create
     user = find_user
```
Comment (chalchuck) [inline app/controllers/sessions_controller.rb]: either move this to a `before_action` or write a better function name and call it directly without assigning to the `user` variable.

{Remind to touch on this}

File: app/controllers/sessions_controller.rb [hunk 9]
```diff
@@ -6,7 +6,7 @@ def create
...
     if user&.authenticate(params[:session][:password])
```
Comment (chalchuck) [inline app/controllers/sessions_controller.rb]: this method has more than two responsibilities, please look into it. 

{Remind to talk on this one too.}

**Review by chalchuck — CHANGES_REQUESTED** [review]
review, fix and revert

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (chalchuck) [reply app/controllers/users_controller.rb:39]: ensure that there is always one admin in the system.

File: app/models/user.rb [hunk 10]
```diff
@@ -14,4 +15,17 @@ def User.digest(string)
                BCrypt::Engine.cost
     BCrypt::Password.create(string, cost: cost)
   end
+
+
+  private
+
+  def admin_exist
```
Comment (chalchuck) [inline app/models/user.rb]: How are you using this method?

Comment (chalchuck) [reply app/models/user.rb]: and the logic says that; if an admin is found; then an error is added to the admin field, is this what you wanted to achieve?

File: app/models/user.rb [hunk 11]
```diff
@@ -1,5 +1,6 @@
 class User < ApplicationRecord
-  before_save { email.downcase! }
+  before_save :format_email
+  # validate :admin_exist, on: [:create, :update]
```
Comment (chalchuck) [inline app/models/user.rb]: to use this method, you probably need to implement a callback that guards the destroy action.

File: app/controllers/users_controller.rb [hunk 12]
```diff
@@ -37,14 +37,18 @@ def update
   end
 
   def destroy
-    User.find(params[:id]).destroy
+    find_user.destroy
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: it is not recommended to use the finder like you are using it here;

Reasons being:
- What happens if `find_user` returns NIL? (doing` nil.destroy` will raise a very nasty error ..)

in the find_user method definition,

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -37,14 +37,18 @@ def update
...
     flash[:success] = "User deleted"
     redirect_to users_url
   end
 
 
   private
 
+  def find_user
+    User.find(params[:id])
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: use the `||=` to assign an @user to this finder, and then rescue from ActiveRecord::RecordNotFound, for cases when the `params[:id]` do not yield an @user object.

File: app/controllers/users_controller.rb [hunk 14]
```diff
@@ -12,7 +12,7 @@ def new
   end
 
   def show
-    @user = User.find(params[:id])
+    @user = find_user
```
Comment (chalchuck) [inline app/controllers/users_controller.rb]: please add ALL the actions that make use of `find_user` to the `before_action` call, this way; the variable `@user `will be available in all the actions that you have specified in the `before_action` call

File: app/controllers/users_controller.rb [hunk 15]
```diff
@@ -16,10 +25,52 @@ def create
...
+
+  def logged_in_user
+    store_location
+    unless logged_in?
+      flash[:danger] = "Please log in."
+      redirect_to login_url
+    end
+  end
+
+  def correct_user
```
Comment (chalchuck) [inline app/controllers/users_controller.rb:66]: Add this method to the `before_action` and this is a question method, please have `?` at the end of the method definition

**Review by chalchuck — CHANGES_REQUESTED** [review]
I have reviwed the PR, please  make the changes herein and submit.

File: app/controllers/users_controller.rb (hunk 15 above)
Comment (llabake) [reply app/controllers/users_controller.rb:66]: It is in the `before action`

**Review by chalchuck — APPROVED** [review]

**Pull request merged**